### PR TITLE
[spirv] Translation of several intrinsic functions

### PR DESCRIPTION
**Pull request by ehsannas**
Translation for the following intrinsic functions:

* Matrix transpose
* smoothstep
* refract
* distance
* isinf
* isnan
* isfinite
* sincos
* saturate
* atan2
* fma
* fmod
* frac
* firstbithigh
* firstbitlow

### CODE REVIEW

**Comment by ehsannas** [comment]
PTAL @antiagainst

**Comment by antiagainst** [comment]
Also needs to rebase

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 1]
```diff
@@ -3116,21 +3130,79 @@ uint32_t SPIRVEmitter::processIntrinsicCallExpr(const CallExpr *callExpr) {
...
+    const auto isNotInf =
+        theBuilder.createUnaryOp(spv::Op::OpLogicalNot, returnType, isInf);
+    return theBuilder.createBinaryOp(spv::Op::OpLogicalAnd, returnType,
+                                     isNotNan, isNotInf);
+  }
+  case hlsl::IntrinsicOp::IOP_sincos: {
+    // Since there is no sincos equivalent in SPIR-V, we need to perform Sin
+    // once and Cos once. We can reuse existing Sine/Cosine handling by creating
+    // a CallExpr for each.
+    CallExpr *sinExpr =
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Oops :-) Done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 2]
```diff
@@ -3116,21 +3130,79 @@ uint32_t SPIRVEmitter::processIntrinsicCallExpr(const CallExpr *callExpr) {
                                            GLSLstd450::GLSLstd450SSign,
                                            /*actPerRowForMatrices*/ true);
   }
+  case hlsl::IntrinsicOp::IOP_isfinite: {
+    // Since OpIsFinite needs the Kernel capability, translation is instead done
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 3]
```diff
@@ -3116,21 +3130,79 @@ uint32_t SPIRVEmitter::processIntrinsicCallExpr(const CallExpr *callExpr) {
...
+    // using OpIsNan and OpIsInf:
+    // isFinite = !isNan && !isInf.
+    const auto arg = doExpr(callExpr->getArg(0));
+    const auto returnType = typeTranslator.translateType(callExpr->getType());
+    const auto isNan =
+        theBuilder.createUnaryOp(spv::Op::OpIsNan, returnType, arg);
+    const auto isInf =
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

**Pull request merged**